TestFoundation/TestUnitVolume.swift

```diff
@@ -0,0 +1,76 @@
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2019 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See http://swift.org/LICENSE.txt for license information
+// See http://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+
+class TestUnitVolume: XCTestCase {
+    func testMetricVolumeConversions() {
+        let cubicKilometers = Measurement(value: 4, unit: UnitVolume.cubicKilometers)
+        XCTAssertEqual(cubicKilometers, Measurement(value: 4e9, unit: UnitVolume.cubicMeters), "Conversion from cubicKilometers to cubicMeters")
+
+        let megaliters = Measurement(value: 3, unit: UnitVolume.megaliters)
+        XCTAssertEqual(megaliters, Measurement(value: 3_000_000, unit: UnitVolume.liters), "Conversion from megaliters to liters")
+
+        let kiloliters = Measurement(value: 2, unit: UnitVolume.kiloliters)
+        XCTAssertEqual(kiloliters, Measurement(value: 2000, unit: UnitVolume.liters), "Conversion from kiloliters to liters")
+
+        let cubicMeters = Measurement(value: 2, unit: UnitVolume.cubicMeters)
+        XCTAssertEqual(cubicMeters, Measurement(value: 2000, unit: UnitVolume.liters), "Conversion from cubicMeters to liters")
+        XCTAssertEqual(kiloliters, cubicMeters, "Conversion from kiloliters to cubicMeters")
+
+        let liters = Measurement(value: 5, unit: UnitVolume.liters)
+        XCTAssertEqual(liters, Measurement(value: 5, unit: UnitVolume.cubicDecimeters), "Conversion from liters to cubicDecimeters")
+        XCTAssertEqual(liters, Measurement(value: 50, unit: UnitVolume.deciliters), "Conversion from liters to deciliters")
+        XCTAssertEqual(liters, Measurement(value: 500, unit: UnitVolume.centiliters), "Conversion from liters to centiliters")
+        XCTAssertEqual(liters, Measurement(value: 5000, unit: UnitVolume.milliliters), "Conversion from liters to milliliters")
+        XCTAssertEqual(liters, Measurement(value: 5000, unit: UnitVolume.cubicCentimeters), "Conversion from liters to cubicCentimeters")
+        XCTAssertEqual(liters, Measurement(value: 5e6, unit: UnitVolume.cubicMillimeters), "Conversion from liters to cubicMillimeters")
+    }
+
+    func testMetricToImperialVolumeConversion() {
+        let liters = Measurement(value: 10, unit: UnitVolume.liters)
+        XCTAssertEqual(liters.converted(to: .cubicInches).value, 610.236, accuracy: 0.001, "Conversion from liters to cubicInches")
+    }
+
+    func testImperialVolumeConversions() {
+        let cubicMiles = Measurement(value: 1, unit: UnitVolume.cubicMiles)
+        XCTAssertEqual(cubicMiles.converted(to: .cubicYards).value, 1760 * 1760 * 1760, accuracy: 1_000_000, "Conversion from cubicMiles to cubicYards")
```

Comment:
Admittedly, this assertion with the insanely high accuracy threshold is kind of ridiculous. I wanted the test to be intuitive, so for people who know that 1 mile = 1760 yards, it makes sense that 1 cubic mile = 1760^3 cubic yards.

Unfortunately, that's not really the case with `Measurement` (even after factoring in floating-point rounding errors) because the conversion factors hardcoded by Foundation often don't use the full range offered by `Double`. In this instance, `cubicMiles` is defined like this (in `Unit.swift`):

```swift
static let cubicMiles           = 4.168e+12 // in liters
```

Whereas the "accurate" conversion factor is more like `4.16818182544058e12` (= 16093.44^3; 1 mile is 16093.44 decimeters) (I think), which leads to the huge discrepancy (in absolute terms) between the actual and expected values.

This inaccuracy is not specific to corelibs-foundation, but is the same in Darwin-Foundation. Out of interest, was it a conscious decision to limit conversion factors to x significant digits or something when the Measurement types were introduced?

Reply:
As far as scl-foundation is concerned it just tries to match Darwin's version as closely as possible, I don't know why Foundation has such low accuracy for its conversions.

The tests should hopefully catch any accidental changes in the future so thanks for adding them, Ive kicked off a CI test for them on Linux and will merge if the tests work ok.